### ISSUE TITLE
Wfs is rounding the output decimals to 6 numbers

### ISSUE THREAD

**Issue by pascal-codetaal** [MapServer/MapServer#6208]
We have a postgres database with some geom data, when importing the wfs to qgis we can see a small offset.

![alt text](https://i.ibb.co/dWgW1PW/Screenshot-2021-01-18-at-13-59-54.png)

This is the geom in the postgres database
`SRID=4326;MULTIPOLYGON(((4.346174668764422 51.0581301240931,0 4.34616114951401,51.05812865886134 0,4.346153577981363 51.05815530900463,0 4.346167288348252)))`

When calling the wfs the decimals are rounded and truncated to 6 decimals.
`<gml:posList srsDimension="2">51.058130 4.346175 51.058129 4.346161 51.058155 4.346154 51.058157 4.346167 51.058130 4.346175 </gml:posList>`

When we ask for the output OGRGML we can pass the LCO:COORDINATE_PRECISION and those are correct.
But with the default wfs we aren't able to change the precision

so following map we are using
```md
MAP

    OUTPUTFORMAT
        NAME "OGRGML"
        DRIVER "OGR/GML"
        FORMATOPTION "LCO:COORDINATE_PRECISION=10"
      END

    OUTPUTFORMAT
        NAME "wfs"
        DRIVER "OGR/GML"
        MIMETYPE "application/json; subtype=geojson; charset=utf-8"
        FORMATOPTION "STORAGE=stream"
        FORMATOPTION "FORM=SIMPLE"
        FORMATOPTION "USE_FEATUREID=true"
        FORMATOPTION "LCO:COORDINATE_PRECISION=15"
    END

    LAYER
        NAME "begraafplaatsen_qgis3__dd84521e_f3b2_4763_9d85_b5a6b41ad756"
        STATUS ON
        TYPE polygon
        EXTENT 4.346118494380637 51.057783128543726 4.34689792795098 51.058230468179445
        METADATA
          "ows_title" "begraafplaatsen_qgis3__dd84521e_f3b2_4763_9d85_b5a6b41ad756"
          "ows_abstract" "begraafplaatsen_qgis3__dd84521e_f3b2_4763_9d85_b5a6b41ad756"
          "wfs_getfeature_formatlist" "wfs, ogrgml"
          "gml_featureid" "id"
          "gml_include_items" "all"
          "gml_types" "auto"
          "wfs_srs" "EPSG:4326" 
        END
        CONNECTIONTYPE POSTGIS
        CONNECTION "host=**** port=**** dbname=geometries user=admin password=*****"
        DATA "geom from begraafplaatsen_qgis3__dd84521e_f3b2_4763_9d85_b5a6b41ad756"
        CLASS
            STYLE
                COLOR 255 0 0 
            END
        END
    END
    WEB
        MINSCALE 10
        MAXSCALE 1550000
        METADATA
            "wms_title" "begraafplaatsen_qgis3__dd84521e_f3b2_4763_9d85_b5a6b41ad756"
            "wms_onlineresource"  "http://localhost:3004/?map=/etc/mapserver/begraafplaatsen_qgis3__dd84521e_f3b2_4763_9d85_b5a6b41ad756.map&service=WMS"
            "wms_srs" "EPSG:4326 EPSG:900913 EPSG:3857 EPSG:31370" 
            "wms_enable_request" "*"
            "wms_feature_info_mime_type" "text/html"
            "wms_format" "image/png"
            "wfs_title" "begraafplaatsen_qgis3__dd84521e_f3b2_4763_9d85_b5a6b41ad756"
            "wfs_onlineresource"  "http://localhost:3004/?map=/etc/mapserver/begraafplaatsen_qgis3__dd84521e_f3b2_4763_9d85_b5a6b41ad756.map&service=WFS"
            "wfs_srs" "EPSG:4326 EPSG:900913 EPSG:3857 EPSG:31370" 
            "wfs_enable_request" "*"
            "wfs_precision" "10"
            "wfs_feature_info_mime_type" "text/html"
        END
         
    END 
END
```

**Comment by jratike80** [comment]
Normally there is no need to configure any outputformat for WFS because GML is the default and supported automatically.  However, I could not find a way to add more decimals into the default GML format and I can confirm your finding that coordinates area rounded to 6 decimals `-78.277153`.

**Comment by pascal-codetaal** [comment]
@jratike80 Do you have any ideas where we could change those values in the code?
Perhaps via a configuration file or global setting file?
We are willing to invest some days to create a pull request.

**Comment by jratike80** [comment]
Sorry, I am a somewhat experienced user but code is not really by business. But perhaps it is somewhere here https://github.com/MapServer/MapServer/blob/main/mapgml.c. With my capabilities I do not see where the coordinates from "shapes" gets truncated.

**Comment by jmckenna** [comment]
I believe one place these 6 decimals are hardcoded in mapgml.c is in the function _gmlWriteBounds_GML2_ around[ line#67](https://github.com/MapServer/MapServer/blob/main/mapgml.c#L67)
```
  msIO_fprintf(stream, "%.6f,%.6f %.6f,%.6f", rect->minx, rect->miny, rect->maxx, rect->maxy );
```
This is also set for the function _gmlWriteBounds_GML3_ around [line#94](https://github.com/MapServer/MapServer/blob/main/mapgml.c#L94)
```
  msIO_fprintf(stream, "%s\t\t<gml:lowerCorner>%.6f %.6f</gml:lowerCorner>\n", tab, rect->minx, rect->miny);
  msIO_fprintf(stream, "%s\t\t<gml:upperCorner>%.6f %.6f</gml:upperCorner>\n", tab, rect->maxx, rect->maxy);
```

I haven't tested what happens when modifying "%.6f" to something like %.12f"   yet

**Comment by pascal-codetaal** [comment]
@jmckenna I think those are for the bounds and not for the polygons/points.lines
```
<wfs:boundedBy>
        <gml:Envelope srsName="urn:ogc:def:crs:EPSG::4326">
            <gml:lowerCorner>51.058060 4.346118</gml:lowerCorner>
            <gml:upperCorner>51.058193 4.346502</gml:upperCorner>
        </gml:Envelope>
    </wfs:boundedBy>
````
```
<wfs:member>
        <ms:begraafplaatsen_qgis3__dd84521e_f3b2_4763_9d85_b5a6b41ad756 gml:id="begraafplaatsen_qgis3__dd84521e_f3b2_4763_9d85_b5a6b41ad756.72">
            <gml:boundedBy>
                <gml:Envelope srsName="urn:ogc:def:crs:EPSG::4326">
                    <gml:lowerCorner>51.058111 4.346346</gml:lowerCorner>
                    <gml:upperCorner>51.058135 4.346367</gml:upperCorner>
                </gml:Envelope>
            </gml:boundedBy>
            <ms:msGeometry>
                <gml:Polygon gml:id="begraafplaatsen_qgis3__dd84521e_f3b2_4763_9d85_b5a6b41ad756.72.1" srsName="urn:ogc:def:crs:EPSG::4326">
                    <gml:exterior>
                        <gml:LinearRing>
                            <gml:posList srsDimension="2">51.058134 4.346346 51.058135 4.346361 51.058112 4.346367 51.058111 4.346353 51.058134 4.346346 </gml:posList>
                        </gml:LinearRing>
                    </gml:exterior>
                </gml:Polygon>
            </ms:msGeometry>
            <ms:id>72</ms:id>
        </ms:begraafplaatsen_qgis3__dd84521e_f3b2_4763_9d85_b5a6b41ad756>
    </wfs:member>
```

To my understanding the writing of the poslist is found in the functions gmlWriteGeometry_GML3 and gmlWriteGeometry_GML2 but i can't see any formatting applied there.

**Comment by geographika** [comment]
Does the LCO work with GeoJSON?

```
OUTPUTFORMAT
    NAME "GeoJSON"
    DRIVER "OGR/GEOJSON"
    MIMETYPE "application/json; subtype=geojson; charset=utf-8"
    FORMATOPTION "STORAGE=stream"
    FORMATOPTION "FORM=SIMPLE"
    FORMATOPTION "COORDINATE_PRECISION=10"
END
```

**Comment by jratike80** [comment]
Yes it works. With `FORMATOPTION "LCO:COORDINATE_PRECISION=20"` 

`"geometry": { "type": "Polygon", "coordinates": [ [ [ -95.07169299999999623196,` 

LCO:COORDINATE_PRECISION=10 yields ` -95.071693`. Should I count "two for integer part, 6 for decimals, and 1 for the sign"?

**Comment by geographika** [comment]
Ah ok. 6 is the default precision for floating point numbers. I'd imagine these are written in the function here (depending on the geometry type):

https://github.com/MapServer/MapServer/blob/20b7671f788952c6097ae154b667dd564cf70764/mapgml.c#L245

If finer control of output is needed then why not use the `OGRGML`?

**Comment by jratike80** [comment]
Is there any easy way to make OGRGML as default format for WFS? WFS clients do not usually include the optional &outputformat= parameter because GML is the default.

**Comment by geographika** [comment]
If setting `&outputformat=ogrgml` in the client-application isn't possible then I'd probably set it in the web server using URL rewriting to add it if missing for WFS requests.

**Comment by jratike80** [comment]
And I would say that it is unnecessarily complicated even I know how to do it and I could perhaps even manage to explain it for the users on the mailing list.

Fortunately the precision is not big issue for most users because we do not do WFS-T, and six decimals is well enough for metric coordinates. Probably that's why it has not been asked before.

BTW Geoserver admin can control the precision this way, the default is unlimited

![image](https://user-images.githubusercontent.com/1751612/105027638-ae21ba80-5a58-11eb-98c7-0a1f0d97d1c2.png)

**Comment by pascal-codetaal** [comment]
Are there any downsides to change the following

```
 msIO_fprintf(stream, "%f,%f,%f ", shape->line[i].point[j].x, shape->line[i].point[j].y, 
```

to
```
 msIO_fprintf(stream, "%0.10f,%0.10f,%0.10f ", shape->line[i].point[j].x, shape->line[i].point[j].y, 

```

**Comment by jratike80** [comment]
What we have now https://mapserver.gis.umn.edu/ogc/wfs_server.html#wfs-server

> gml_[item name]_precision
(Optional) Specifies the precision of the indicated field for formats where this is significant, such as Shapefiles. Precision is the number of decimal places, and is only needed for “Real” fields. Currently this is only used for OGR based output formats, not the WFS GML2/GML3 output.

As a user I think I would easily learn to use a keyword

`gml_geometry_precision`

Alternatively `wfs_geometry_precision` would have the same effect on all outputformats.

**Comment by pascal-codetaal** [comment]
@jratike80 Indeed we did play with those values and they seem to work on all items except geom.

**Comment by pascal-codetaal** [comment]
Where are those values parsed in the code? 
I can't seem to find anything that maps to _precision

**Comment by jratike80** [comment]
Somewhere that deals with OGR formats.

> Currently this is only used for OGR based output formats, not the WFS GML2/GML3 output.

**Comment by jmckenna** [comment]
@jratike80 I confirmed yesterday that 'gml_geometry_precision' does not work for WFS.

**Comment by jmckenna** [comment]
This is a longtime user request/report, this hardcoded 6 decimal points for WFS (most reports with PostgreSQL, SQLite). I think we should work on a pull request to make sure that the layer metadata 'gml_geometry_precision' and 'wfs_geometry_precision' are enabled for these cases.  @pascal-codetaal are you offering to tackle this? :)

**Comment by jmckenna** [comment]
By the way, this has indeed been asked before, the request goes way back.  I feel it is time to solve it :)  Great report @pascal-codetaal

**Comment by pascal-codetaal** [comment]
@jmckenna Me and @jaspervercnocke will try to tackle this issue.
Are there any guidelines we should follow?
We already changed some code locally and do get the wanted result. 

So you would like us to implement following metadata?

```
MAP
    STATUS ON
    PROJECTION  
        "init=epsg:4326"
    END
    LAYER
        NAME "test_layer"
        STATUS ON
        TYPE polygon
        EXTENT 4.346118494380637 51.057783128543726 4.34689792795098 51.058230468179445
        METADATA
          "ows_title" "test layer"
          "ows_abstract" "test layer"
          "gml_featureid" "id"
          "gml_include_items" "all"
          "gml_types" "auto"
          "gml_geometry_precision" "10"
          "wfs_srs" "EPSG:4326" 
        END
        CONNECTIONTYPE POSTGIS
        CONNECTION "host=postgres port=5432 dbname=****** user=**** password=*********"
        DATA "geom from test_layer"
    END
    WEB
        METADATA
            "wfs_title" "test layer"
            "wfs_onlineresource"  "http://localhost/?map=/etc/mapserver/test_layer.map&service=WFS"
            "wfs_srs" "EPSG:4326 EPSG:900913 EPSG:3857 EPSG:31370" 
            "wfs_enable_request" "*"
            "wfs_geometry_precision" "10"
        END
    END 
END
```

Which branch should we start from?

**Comment by jmckenna** [comment]
Please start from the default 'main' branch.  Yes your mapfile example is exact (enabling "wfs_geometry_precision").  

@pascal-codetaal I really appreciate that you and @jaspervercnocke are offering to jump on this.  Users have been reporting this for Oracle, Postgres, Sqlite, for years.  I have a test case locally (with a spatialite db) to help test, please keep me posted.  And thank-you both!!

**Comment by jmckenna** [comment]
Also, please as part of this, invest time to make sure that msautotest runs fine after your changes (there is a /msautotest/wxs/ folder in the source that contains WFS tests).  msautotest can be setup locally on your machines.  Often the time to learn and make sure msautotest runs, takes about the same or more time than to add your source code changes :) So please add that to the list, as it is very important.  Thanks again!

**Comment by jratike80** [comment]
There is now one setting in WEB - METADATA and another in LAYER - METADATA. Keywords are gml_geometry_precision and wfs_geometry_precision.

Could it be this way:
- without keywords the old default applies
- WEB METADATA will be the default for all layers
- LAYER METADATA overrides the default and what is set in WEB METADATA

I invented `wfs_geometry_precision` to allow to adjust GeoJSON and other possible extra formats by the same. Item `gml_geometry_precision` should not do that or it would be misleading. But do we need `wfs_geometry_precision` because other formats than the default GML must anyway have an OUTPUTFORMAT block and precision can be set there - though that affects the whole service and layer level overrides would not be possible. Different precisions could be useful if there are both geographic and projected layers in the service. But it may also lead to more confusion if precision is set both in METADATA and within the OUTPUTFORMAT.
(Maybe I do not dare to think if somebody would like to have different precision by the srsName used in the GetFeature request).

**Comment by jmckenna** [comment]
My personal opinion is that yes we need "wfs_geometry_precision" (working for Oracle, Postgres, Sqlite data, which it currently does not) at the layer level, as most users will not have an OUTPUTFORMAT block.

**Comment by jmckenna** [comment]
> Could it be this way:
>     * without keywords the old default applies
>     * WEB METADATA will be the default for all layers
>     * LAYER METADATA overrides the default and what is set in WEB METADATA

+1

**Comment by pascal-codetaal** [comment]
I think we have a first working version with the web metadata -> wfs_geometry_precision.
Tomorrow we will create a pull request.

**Comment by jmckenna** [comment]
thanks.  As you work on this pull request please include a new test for this metadata in /msautotest/wxs/ as part of your pull request ( follow https://mapserver.org/development/tests/autotest.html ).  thanks!

**Comment by pascal-codetaal** [comment]
Just a question for the tests.
Should we create a separate test our pull request?
Or should we change/adapt some existing tests?

**Comment by jmckenna** [comment]
You can include the test as part of your pull request.  

You can decide (once you spend a few days setting up the tests and learning msautotest) whether your test fits into an existing test, or whether you need a new test.  Take your time (I found that setting up msautotest properly to understand it, took me many days last time, so, take your time to learn msautotest and then adapt/create new test).  

Short answer: it's up to your preference.

**Comment by geographika** [comment]
@pascal-codetaal - it may also be worth posting to the mapserver-dev mailing list. 

As this change will only affect the GML output it may be more coherent to use `gml_[geometry name]_precision` similar to other GML settings documented at https://mapserver.org/ogc/wfs_server.html#layer-object
This could clash with `gml_[item name]_precision`, but the geometry name shouldn't be a attribute and geometry field. 
A `METATADATA` section would look similar to:
```
  METADATA
    "wfs_title"         "multilinestring"
    "wfs_description"   "multilinestring"
    "wfs_featureid"     "id"
    "gml_geometries"    "msGeometry"
    "gml_msGeometry_type" "multiline"
    "gml_msGeometry_precision" "10"
  END
```

**Comment by jmckenna** [comment]
I think we are all speaking of "gml_[geometry name]_precision" and "wfs_[geometry name]_precision", at least that's what I was understanding (following the existing functionality for other formats for those metadata, except now we will finally enable them for WFS users of Oracle, PostgreSQL, SQLite...    Currently, as reported here, those existing metadata settings do not apply to those WFS source formats.

**Comment by jmckenna** [comment]
Request sent to -dev list regarding feedback for this change ( https://lists.osgeo.org/pipermail/mapserver-dev/2021-January/016344.html ).

**Comment by jmckenna** [comment]
@pascal-codetaal @jaspervercnocke thanks for this effort, and for including several news tests for precision in msautotest.

I've tested it locally using your Orbit-GT-Belgium fork and 'wfs_geometry_precision' branch.  Wow works great! (testing on Windows)  I've now tested with a Spatialite database, and latest PostgreSQL/PostGIS database.  Without your changes, if I set `wfs_geometry_precision` at the web or layer levels, the coordinates returned through WFS are defaulted to 6 precision, but after your change I am able to set something like `'wfs_geometry_precision' '20'` and the resulting GetFeature response contains geometry with that precision (as below).  I've also verified that setting it at the LAYER level overrides the same setting at the WEB level (as one of your new tests confirm as well).
```
<ms:msGeometry>
<gml:Polygon gml:id=".1" srsName="urn:ogc:def:crs:EPSG::4326">
<gml:exterior>
<gml:LinearRing>
<gml:posList srsDimension="2">
-28.95559661226171144222 28.97826256685724288786 -29.25738697684625577722 29.32516645683259071120 -29.74376555757736895202 29.01841515474802690733 -30.07005055106825608391 28.84839969250774061038 -30.22621672945430049140 28.29106937023991008573 -30.54573211031495105772 28.10720462414542453189 -30.64510588961222481430 27.74939700695648525652 -29.87595387137998415028 26.99926191580763656930 -29.24271087007535996349 27.53251102062747790455 -28.85146860119358791508 28.07433841320778356021 -28.64750172293756946829 28.54170006685549765280 -28.95559661226171144222 28.97826256685724288786
</gml:posList>
</gml:LinearRing>
</gml:exterior>
</gml:Polygon>
</ms:msGeometry>
```

This is a great addition (and one that has been reported for Oracle, PostGIS, Sqlite users for a very long time).  Let's give some more days for others to test this locally & provide feedback.

Thank-you again for putting in this effort. 🙏

**Comment by pascal-codetaal** [comment]
We were happy to help out, we are creating a new geo platform and one of the building blocks is mapserver.
So we should thank the people building mapserver.
And if you would like to thank someone it should be our employee [orbit-gt](https://github.com/Orbit-GT-Belgium/).

**Comment by jmckenna** [comment]
Thanks again @pascal-codetaal @jaspervercnocke, your changes will be included in the upcoming 8.0 release (scheduled for 2nd April).  I've made a note of your changes in the release plan (https://github.com/MapServer/MapServer/wiki/MapServer-8.0-Release-Plan).
Please keep the changes coming, as it's great to have new life into the project.
